overlord/configstate/configcore/picfg.go

```diff
@@ -89,16 +90,31 @@ func updatePiConfig(path string, config map[string]string) error {
 	return nil
 }
 
-func piConfigFile(opts *fsOnlyContext) string {
+func piConfigFile(opts *fsOnlyContext) (string, error) {
 	rootDir := dirs.GlobalRootDir
+	subdir := "/boot/uboot"
 	if opts != nil {
 		rootDir = opts.RootDir
+	} else {
+		// not a filesystem only apply, so we may be operating on a run system
+		// on UC20, in which case we shouldn't use the /boot/uboot/ option and
+		// instead should use /run/mnt/ubuntu-seed/
+		mode, _, _ := boot.ModeAndRecoverySystemFromKernelCommandLine()
+		// TODO: what about recover mode?
```

Comment:
the directory would also work in install and recover, no?  any reason not to just let it be set?  we either needs to do that or error, because as is we will just write to the wrong place?

Reply:
hmm, yes that directory would also work for install mode and recover, but what's the use case for setting config from install or recover mode? It feels unexpected to me that if you:
* seed a UC20 device with some defaults from your gadget for pi-config
* those get applied and are used in run mode
* later, you decide to change a pi-config which takes effect on run mode and reboot
* now for some reason you end up in recover mode
* recover mode seeding then overwrites the new setting you manually set for the device
* you go back to run mode, now the value of pi-config in state doesn't match what's in config.txt

Really this harkens back to the fact that in recover mode we don't have a good answer about what to do about gadget pi-config values. I'm not sure we should hard-fail the config here, since that might break seeding, but right now what happens is that we return the wrong file here, but actually applying the config checks if the file exists, so in this case the wrong file won't exist and so we don't process any of the config, which in effect is a silent failure. Perhaps we should make this behavior more explicit by returning a specific error here and have the config check for that error instead

Reply:
I made it so that on install and recover mode we explicitly just error here, and then in the calling function we return without processing any config if we that error returned from this function so that we have expected behavior and we don't attempt to write to the wrong file even if doing so is currently harmless in UC20.